tests/integration_tests.rs

```diff
@@ -0,0 +1,644 @@
+
+    // Run the server with client auth enabled.
+    let mut server_crypto = server_crypto();
+    server_crypto.enable_client_auth(true);
+    let server = Server::run(server_config(server_crypto)?)?;
+
+    // Create the client with an unknown CA.
+    let mut client_crypto = ClientConfig::new().unwrap();
+    client_crypto
+        .set_cert(UNTRUSTED_CLIENT_CERT.chain(), UNTRUSTED_CLIENT_CERT.key())
+        .unwrap();
+    for cert in SERVER_CERT.chain() {
+        client_crypto.add_trusted_cert(cert).unwrap();
+    }
+    let client = Client::connect(client_config(client_crypto), server.addr).await?;
+    let (send, _) = client.conn.open_bi().await?;
+    let request = &b"hello world"[..];
+    let err = Client::send_request(send, request)
+        .await
+        .expect_err("expected connection failure");
+    match err {
+        WriteError::ConnectionLost(ConnectionError::ConnectionClosed(e)) => {
+            assert_eq!(TransportErrorCode::crypto(UNKNOWN_CA), e.error_code);
+            Ok(())
+        }
+        _ => Err(anyhow!("unexpected error: {err}")),
+    }
+}
+
+#[tokio::test]
+async fn zero_rtt_success() -> Result<()> {
+    let _guard = subscribe();
+
+    // Run the server.
+    let server = Server::run(server_config(server_crypto())?)?;
+
+    let cc1 = client_crypto();
+    let session_cache = cc1.get_session_cache();
+    let client1 = Client::connect_no_zero_rtt(client_config(cc1), server.addr).await?;
+    client1.ping_pong().await?;
+
+    info!("initial connection complete");
+
+    // Create the second client with the shared session cache.
+    let mut cc2 = client_crypto();
+    cc2.set_session_cache(session_cache);
+    let (client2, zero_rtt) = Client::connect_zero_rtt(client_config(cc2), server.addr).await?;
+    client2.ping_pong().await?;
+    assert!(zero_rtt.await, "expected 0-RTT");
+    Ok(())
+}
+
+#[tokio::test]
+async fn zero_rtt_rejected() -> Result<()> {
+    let _guard = subscribe();
+
+    // Run the server.
+    let server = Server::run(server_config(server_crypto())?)?;
+
+    let cc1 = client_crypto();
+    let session_cache = cc1.get_session_cache();
+    let client1 = Client::connect_no_zero_rtt(client_config(cc1), server.addr).await?;
+    client1.ping_pong().await?;
+
+    info!("initial connection complete");
+
+    // Start a new server, to ensure that it does not have a session cached for this client. This
+    // will force the server to reject the 0-RTT attempt.
+    let server = Server::run(server_config(server_crypto())?)?;
+
+    // Create the second client with the shared session cache.
+    let mut cc2 = client_crypto();
+    cc2.set_session_cache(session_cache);
+    let (client2, zero_rtt) = Client::connect_zero_rtt(client_config(cc2), server.addr).await?;
+    let (send, _) = client2.conn.open_bi().await?;
+
+    // Buy time for the driver to process the server's NewSessionTicket
+    tokio::time::sleep_until(Instant::now() + Duration::from_millis(100)).await;
```

Comment:
@djc adding this sleep immediately after we open the connection seems to avoid the connection timeout. Thoughts?

Reply:
@Ralith is there a good reason for this?

Reply:
Per [RFC 9001 § Session Resumption](https://www.rfc-editor.org/rfc/rfc9001.html#name-session-resumption), CRYPTO frames bearing a TLS `NewSessionTicket` are delivered *after* the handshake completes, i.e. in 1-RTT data. Meanwhile, this test proceeds to attempt a 0-RTT connection as soon as it's received the 0.5-RTT "goodbye!" message from the server. This delay must be intended to reduce the odds of the connection being dropped before the `NewSessionTicket` is processed.

I'm not actually able to reproduce Quinn's existing test taking 10 seconds at the moment, nor does it fail for me if I remove the timeout. I'm guessing a more robust and straightforward approach would be to wait for some 1-RTT application data before dropping the initial connection.

Reply:
Note also the above explanation isn't consistent with the derived use of sleep_until that this comment thread is attached to, because here it occurs *after* the second connection is established. It'd be interesting to know where either/both tests are getting stuck waiting for a timeout.

Reply:
Thanks @Ralith. Let's just land this for now. Quinn should probably have a copy of these integration tests anyway ... we can defer the investigation until then.

Reply:
SGTM, but maybe delete this misleading comment (replace with a TODO to investigate?).

Reply:
Done. FYI that comment was copied from the quinn test .. so probably should be replaced there as well.

Reply:
Thanks! In the Quinn test the comment is accurate, because the sleep is placed in a different context.